File: 3_airport/employee.cpp
#include <iostream>
#include "employee.h"
#include "planeComponent.h"

using namespace std;

//========================================
//          --- Employee ---
//========================================

// Constructor + Destructor
Employee::Employee():name(randName()){
	worked=0;
	cout << " Employee " << id << " just created!" << endl;
}
Employee::~Employee(){
	cout << " Employee " << id << " just destroyed!" << endl;
}


// Getters + Setter
const String &Employee::getName()const{ return name; }
bool Employee::checkWork()const{ return worked; }
void Employee::workComplete(){ worked=1; }


// toString virtual implementation for Employee class and sub-classes
String Employee::toString()const{
	String tmp;
	// Pass the object id
	tmp.concat(">Id= ");
	tmp.concat(int_to_String(id));

	// Pass the employee's name
	tmp.concat(" Name= ");
	tmp.concat(name);

	// And his work status
	tmp.concat(" Work= ");
	tmp.concat(bool_to_String(worked));

	return tmp;
}
bool Employee::equal(const Object *obj)const{
	if(id!=obj->getId()){ return 0; }
	if( !(name.equal(((Employee *)obj)->getNamePtr())) ){ return 0; }
	if( worked!=((Employee *)obj)->checkWork()){ return 0; }
	return 1;
}


//========================================
//       --- SecurityEmployee ---
//========================================

// Constructors + Destructor
SecurityEmployee::SecurityEmployee(){
	cout << " SecurityEmployee " << id << " just created!" << endl;
}
SecurityEmployee::SecurityEmployee(const SecurityEmployee& obj){
	id=obj.getId();
	name=obj.getName();
	worked=obj.checkWork();
	cout << " SecurityEmployee " << id << " just created!" << endl;
}
SecurityEmployee::~SecurityEmployee(){
	cout << " SecurityEmployee " << id << " just destroyed!" << endl;
}

// Clone functions
Object *SecurityEmployee::clone(){
	SecurityEmployee *obj=new SecurityEmployee(*this);
	return obj;
}

// Employee virtuals
void SecurityEmployee::workOn(PlaneComponent& cmp){
	cout << " Security Employee checking plane compartment: ";
	print(cmp.getDesc());
	cmp.setSec(1);
}
void SecurityEmployee::report()const{
	if(worked){cout << " Security Check: OK" << endl;}
	else{cout << " Security Check: NOT OK" << endl;}
}


//========================================
//      --- MaintenanceEmployee ---
//========================================

// Constructors + Destructor
MaintenanceEmployee::MaintenanceEmployee(){
	cout << " MaintenanceEmployee " << id << " just created!" << endl;
}
MaintenanceEmployee::MaintenanceEmployee(const MaintenanceEmployee& obj){
	id=obj.getId();
	name=obj.getName();
	worked=obj.checkWork();
	cout << " MaintenanceEmployee " << id << " just created!" << endl;
}
MaintenanceEmployee::~MaintenanceEmployee(){
	cout << " MaintenanceEmployee " << id << " just destroyed!" << endl;
}

// Clone functions
Object *MaintenanceEmployee::clone(){
	MaintenanceEmployee *obj=new MaintenanceEmployee(*this);
	return obj;
}

// Employee virtuals
void MaintenanceEmployee::workOn(PlaneComponent& cmp){
	cout << " Maintenance Employee checking plane compartment: ";
	print(cmp.getDesc());
	cmp.setMnt(1);
}
void MaintenanceEmployee::report()const{
	if(worked){cout << " Maintenance Check: OK" << endl;}
	else{cout << " Maintenance Check: NOT OK" << endl;}
}


//========================================
//       --- CleaningEmployee ---
//========================================

// Constructors + Destructor
CleaningEmployee::CleaningEmployee(){
	cout << " CleaningEmployee " << id << " just created!" << endl;
}
CleaningEmployee::CleaningEmployee(const CleaningEmployee& obj){
	id=obj.getId();
	name=obj.getName();
	worked=obj.checkWork();
	cout << " CleaningEmployee " << id << " just created!" << endl;
}
CleaningEmployee::~CleaningEmployee(){
	cout << " CleaningEmployee " << id << " just destroyed!" << endl;
}

// Clone functions
Object *CleaningEmployee::clone(){
	CleaningEmployee *obj=new CleaningEmployee(*this);
	return obj;
}

// Employee virtuals
void CleaningEmployee::workOn(PlaneComponent& cmp){
	cout << " Cleaning Employee checking plane compartment: ";
	print(cmp.getDesc());
	cmp.setCln(1);
}
void CleaningEmployee::report()const{
	if(worked){cout << " Cleaning: OK" << endl;}
	else{cout << " Cleaning: NOT OK" << endl;}
}

File: 4_airportJava/Employee.java
package oopjava;

import java.util.Random;

public abstract class Employee extends Object{
        
        protected int id;           // The employee's unique id
	protected String name;      // The employee's name
	protected boolean worked;   // The employee's work state
        private static int idCount; // A counter to ensure unique id's are given
        private static Random rnd;
        static{ rnd = new Random(); }
        static{ idCount=0; }    // Initialize the static id Counter
        
        private static String[] Names={"izambo", "errikos", "spyros", "nikosPerdikopanis", 
        "alexkalts", "vaggelisd", "pstamat", "harry", "stef", "sdi1800164"};
        
        private static String randName(){ return Names[rnd.nextInt(10)]; }
        
        // Getters / Setter
        public int getId(){ return id; }
	public String getName(){ return name; }
	public boolean checkWork(){ return worked; }
	public void workComplete(){ worked=true;  } 
    
        public Employee(){
            id=idCount++;
            name=new String();
            name=randName();
            worked=false;
        }

        @Override
	public abstract Object clone();
        @Override
	public String toString(){
            String tmp=new String();
            // Pass the object id
            tmp=tmp.concat(">Id= ");
            tmp=tmp.concat(String.valueOf(id));

            // Pass the employee's name
            tmp=tmp.concat(" Name= ");
            tmp=tmp.concat(name);

            // And his work status
            tmp=tmp.concat(" Work= ");
            tmp=tmp.concat(String.valueOf(worked));
            return tmp;
        }
        @Override
        public boolean equals(Object obj){
            if( id!=((Employee)obj).getId()){ return false; }
            if( !(name.equals(((Employee)obj).getName()))){ return false; }
            if( worked!=((Employee)obj).checkWork()){ return false; }
            return true;
        };

	public abstract void workOn(PlaneComponent cmp);
	public abstract void report();
}
